File: clients/DotNET/Zbus/Mq/ConsumeThread.cs
using log4net;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Sockets;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace Zbus.Mq
{
    public class ConsumeThread
    {
        private static readonly ILog log = LogManager.GetLogger(typeof(ConsumeThread));

        public Func<MqClient> ClientFactory { get; set; }
        public Action<Message, MqClient> MessageHandler { get; set; }
        public int ThreadCount { get; set; }
        public bool RunInPool { get; set; }
        public int ConsumeTimeout { get; set; }
        public int? ConsumeWindow { get; set; } 

        public string Topic { get; private set; }
        public string Token { get; set; }
        public ConsumeGroup ConsumeGroup { get; private set; } 
         

        private CancellationTokenSource cts = new CancellationTokenSource(); 
        private Thread[] consumeThreadList;
        private MqClient[] clients;

        public ConsumeThread(string topic, ConsumeGroup group = null)
        {
            ConsumeTimeout = 10000; //10s 
            Topic = topic;
            ConsumeGroup = group==null? new ConsumeGroup(topic): group;
            RunInPool = false;
            ThreadCount = 1; 
        }

        private async Task<Message> TakeAsync(MqClient client, CancellationToken? token=null)
        {
            Message res = await client.ConsumeAsync(Topic, ConsumeGroup.GroupName, ConsumeWindow, token);
            if (res == null) return res;
            if(res.Status == "404")
            {
                await client.DeclareGroupAsync(Topic, ConsumeGroup, token);
                return await TakeAsync(client, token);
            }

            if(res.Status == "200")
            {
                return res;
            }
            throw new MqException(res.BodyString); 
        }

        public void Start()
        {
            if (MessageHandler == null)
            {
                throw new InvalidOperationException("Missing MessageHandler");
            }
            if (ClientFactory == null)
            {
                throw new InvalidOperationException("Missing ClientFactory");
            }

            lock (this)
            {
                if (this.consumeThreadList != null) return;
            }
            this.consumeThreadList = new Thread[ThreadCount];
            this.clients = new MqClient[ThreadCount];

            for(int i = 0; i < this.consumeThreadList.Length; i++)
            {
                MqClient client = this.clients[i] = ClientFactory();
                this.consumeThreadList[i] = new Thread(async () =>
                {
                    using (client) { 
                        while (!cts.IsCancellationRequested)
                        {
                            Message msg;
                            try
                            {
                                msg = await TakeAsync(client, cts.Token);
                                if (msg == null) continue;
                                if (RunInPool)
                                {
                                    await Task.Run(() =>
                                    {
                                        MessageHandler(msg, client);
                                    });
                                }
                                else
                                {
                                    MessageHandler(msg, client);
                                }
                            } 
                            catch (Exception e)
                            { 
                                if (e is SocketException || e is IOException)
                                {
                                    client.Dispose();
                                    Thread.Sleep(3000);
                                }  
                                log.Error(e);
                            }
                        }
                    }
                });
            } 
            foreach(Thread thread in this.consumeThreadList)
            {
                thread.Start();
            }
        }

        public void Stop()
        {  
            cts.Cancel();
            for (int i = 0; i < this.clients.Length; i++)
            {
                try
                {
                    this.clients[i].Dispose();
                }
                catch(Exception e)
                {
                    log.Error(e.Message, e);
                    //ignore
                }
            }
        } 
    } 
}

File: src/main/java/io/zbus/mq/ConsumeThread.java
package io.zbus.mq;

import java.io.Closeable;
import java.io.IOException;
import java.nio.channels.ClosedByInterruptException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.TimeUnit;

import io.zbus.kit.ThreadKit.ManualResetEvent;
import io.zbus.kit.logging.Logger;
import io.zbus.kit.logging.LoggerFactory;
import io.zbus.mq.Protocol.ConsumeGroupInfo;

public class ConsumeThread implements Closeable{
	private static final Logger log = LoggerFactory.getLogger(ConsumeThread.class);  
	protected final MqClient client;
	
	protected String topic;  
	protected Integer topicMask;
	protected String token;
	protected ConsumeGroup consumeGroup;
	protected int consumeTimeout = 10000;
	protected Integer consumeWindow;
	
	private String consumeGroupName; //internally upated
	 
	
	protected ExecutorService consumeRunner;
	protected MessageHandler messageHandler;
	
	protected ManualResetEvent active = new ManualResetEvent(true); 
	
	protected RunningThread consumeThread;
	 
	public ConsumeThread(MqClient client, String topic, ConsumeGroup group, Integer topicMask){
		this.client = client;
		this.topic = topic;
		this.consumeGroup = group;
		this.topicMask = topicMask;
	}
	
	public ConsumeThread(MqClient client, String topic){
		this(client, topic, null, null);
	}
	
	public ConsumeThread(MqClient client){
		this(client, null);
	}
	 
	public synchronized void start() {
		start(false);
	}
	
	public synchronized void start(boolean pauseOnStart) {
		if(this.topic == null){
			throw new IllegalStateException("Missing topic");
		}
		if(this.messageHandler == null){
			throw new IllegalStateException("Missing consumeHandler");
		}
		if(pauseOnStart){
			active.reset();
		}
		
		if(this.consumeGroup == null){
			this.consumeGroup = new ConsumeGroup();
			consumeGroup.setGroupName(this.topic);
		}  
		this.client.setToken(token);
		this.client.setInvokeTimeout(consumeTimeout);
		try { 
			ConsumeGroupInfo info = this.client.declareGroup(topic, consumeGroup, topicMask);
			consumeGroupName = info.groupName; //update groupName
		} catch (IOException e) { 
			log.error(e.getMessage(), e);
		} catch (InterruptedException e) { 
			log.error(e.getMessage(), e);
		}
		
		
		consumeThread = new RunningThread();
		consumeThread.start();
	}
	
	public void pause(){
		try {
			client.unconsume(topic, consumeGroupName); //stop consuming in serverside
			consumeThread.running = false;
			consumeThread.interrupt();
			
		} catch (Exception e) {
			log.error(e.getMessage(), e);
		}  
		
		active.reset();
	}
	
	public void resume(){  
		active.set();
		if(consumeThread == null || !consumeThread.running) {
			consumeThread = new RunningThread();
			consumeThread.start();
		}
	}
	
	public Message take() throws IOException, InterruptedException {  
		Message res = null;
		try {  
			res = client.consume(topic, this.consumeGroupName, this.getConsumeWindow()); 
			if (res == null) return res; 
			Integer status = res.getStatus();
			if (status == 404) { 
				ConsumeGroupInfo info = client.declareGroup(topic, consumeGroup, topicMask);
				consumeGroupName = info.groupName; //update groupName
				return take();
			}
			
			if (status == 200) { 
				String originUrl = res.getOriginUrl();
				if(originUrl != null){ 
					res.removeHeader(Protocol.ORIGIN_URL);
					res.setUrl(originUrl);   
					res.setStatus(null);
					String method = res.getOriginMethod();
					if(method != null){
						res.setMethod(method);
						res.removeHeader(Protocol.ORIGIN_METHOD);
					}
					return res;
				}
				
				Integer originStatus = res.getOriginStatus();
				if(originStatus != null){ 
					res.removeHeader(Protocol.ORIGIN_STATUS);  
					res.setStatus(originStatus);
					return res;
				}
				
				res.setStatus(null);//default to request type
				return res;
			}
			
			throw new MqException(res.getBodyString());
		} catch (ClosedByInterruptException e) {
			throw new InterruptedException(e.getMessage());
		}   
	} 
	
	
	class RunningThread extends Thread {
		volatile boolean running = true; 
		
		public void run() { 
			while (running) {
				try { 
					final Message msg;
					try {
						while(running){
							active.await(1000, TimeUnit.MILLISECONDS); 
							if(active.isSignalled()) break;
						}
						if(!running) break; 
						msg = take();
						if(msg == null) continue;
						if(!running) break;
						
						if(messageHandler == null){
							throw new IllegalStateException("Missing ConsumeHandler");
						}
						
						if(consumeRunner == null){
							try{
								messageHandler.handle(msg, client);
							} catch (Exception e) {
								log.error(e.getMessage(), e);
							}
						} else {
							consumeRunner.submit(new Runnable() { 
								@Override
								public void run() {
									try{
										messageHandler.handle(msg, client);
									} catch (Exception e) {
										log.error(e.getMessage(), e);
									}
								}
							});
						}
						
					} catch (InterruptedException e) {
						client.close(); 
						break;
					}  
					
				} catch (IOException e) {  
					log.error(e.getMessage(), e);  
				}
			}
		}
	}
	  

	@Override
	public void close() throws IOException {
		consumeThread.interrupt();  
	} 

	public ExecutorService getConsumeRunner() {
		return consumeRunner;
	}

	public void setConsumeRunner(ExecutorService consumeRunner) {
		this.consumeRunner = consumeRunner;
	}

	public void setConsumeTimeout(int consumeTimeout) {
		this.consumeTimeout = consumeTimeout;
	}

	public void setMessageHandler(MessageHandler messageHandler) {
		this.messageHandler = messageHandler;
	} 
	
	public String getTopic() {
		return topic;
	}

	public void setTopic(String topic) {
		this.topic = topic;
	}  
	
	public String getToken() {
		return token;
	}

	public void setToken(String token) {
		this.token = token;
	}

	public ConsumeGroup getConsumeGroup() {
		return consumeGroup;
	}

	public void setConsumeGroup(ConsumeGroup consumeGroup) {
		this.consumeGroup = consumeGroup;
	}

	public int getConsumeTimeout() {
		return consumeTimeout;
	}

	public MessageHandler getMessageHandler() {
		return messageHandler;
	}

	public Integer getConsumeWindow() {
		return consumeWindow;
	}

	public void setConsumeWindow(Integer consumeWindow) {
		this.consumeWindow = consumeWindow;
	}

	public MqClient getClient() {
		return client;
	}

	public Integer getTopicMask() {
		return topicMask;
	}

	public void setTopicMask(Integer topicMask) {
		this.topicMask = topicMask;
	}  
	
}